dcicutils/sheet_utils.py

```diff
@@ -0,0 +1,679 @@
+                processed_row_data: AnyJsonData = self._process_row(tab_name, state, row_data)
+                sheet_content.append(processed_row_data)
+            self.content_by_tab_name[tab_name] = sheet_content
+        return self.content_by_tab_name
+
+    def parse_cell_value(self, value: SheetCellValue) -> AnyJsonData:
+        return prefer_number(value) if self.prefer_number else value
+
+
+class TableSetManagerRegistry:
+
+    def __init__(self):
+        self.manager_table: Dict[str, Type[AbstractTableSetManager]] = {}
+        self.regexp_mappings: List[Tuple[Regexp, Type[AbstractTableSetManager]]] = []
+
+    def register(self, regexp: Optional[str] = None):
+        def _wrapped_register(class_to_register: Type[AbstractTableSetManager]):
+            if regexp:
+                self.regexp_mappings.append((re.compile(regexp), class_to_register))
+            for ext in class_to_register.ALLOWED_FILE_EXTENSIONS:
+                existing = self.manager_table.get(ext)
+                if existing:
+                    raise Exception(f"Tried to define {class_to_register} to extension {ext},"
+                                    f" but {existing} already claimed that.")
+                self.manager_table[ext] = class_to_register
+            return class_to_register
+        return _wrapped_register
+
+    register1 = register
+
+    def manager_for_filename(self, filename: str) -> Type[AbstractTableSetManager]:
+        base: str = os.path.basename(filename)
+        suffix_parts = base.split('.')[1:]
+        if suffix_parts:
+            for i in range(0, len(suffix_parts)):
+                suffix = f".{'.'.join(suffix_parts[i:])}"
+                found: Optional[Type[AbstractTableSetManager]] = self.manager_table.get(suffix)
+                if found:
+                    return found
+        else:
+            special_case: Optional[Type[AbstractTableSetManager]] = self.manager_for_special_filename(filename)
+            if special_case:
+                return special_case
+        raise LoadArgumentsError(f"Unknown file type: {filename}")
+
+    def manager_for_special_filename(self, filename: str) -> Optional[Type[AbstractTableSetManager]]:
+        for pattern, manager_class in self.regexp_mappings:
+            if pattern.match(filename):
+                return manager_class
+        return None
+
+
+TABLE_SET_MANAGER_REGISTRY = TableSetManagerRegistry()
+
+
+@TABLE_SET_MANAGER_REGISTRY.register()
+class XlsxManager(FlattenedTableSetManager):
+    """
+    This implements the mechanism to get a series of rows out of the sheets in an XLSX file.
+    """
+
+    ALLOWED_FILE_EXTENSIONS = ['.xlsx']
+
+    @classmethod
+    def _all_rows(cls, sheet: Worksheet):
+        row_max = sheet.max_row
+        for row in range(2, row_max + 1):
+            yield row
+
+    @classmethod
+    def _all_cols(cls, sheet: Worksheet):
+        col_max = sheet.max_column
+        for col in range(1, col_max + 1):
+            yield col
```

Comment:
For both of these are starting off the initial index - you should mention why and what the expected structure is. Functions with names like "all" here imply all to me, not necessary "all minus headers" or in the case of columns I'm not really sure what you're cutting off, all depends on expected structure.

Reply:
There is always a header in this format, so this is all of the data columns and rows. If there is not a header, none of this will work. I could rename this to `_all_data_rows` but it will not eliminate the risk, which is not checkable and must simply be documented. We don't have a submission protocol for non-headered files. TableSets, when I document them, must have headers, just as dictionaries cannot have keyless entries.

Reply:
Well, in the case of using this tool for items, we can check that the headings are things in the schema. I don't think it does that now, but it now could. data is unlikely to accidentally match. for tablesets, the lower level abstraction, there is no such reference so if you do:
```
1 2 3
4 5 6
7 8 9
```
you'll just get
```
[
    {'1': '4', '2': '5', '3': '6'},
    {'1': '7', '2': '8', '3': '9'}
]
```
and the effect won't be data loss, "just" data misuse. The first row isn't skipped, it's used as a header.